{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "62926141-6cd8-446f-ae25-fddc655167a3",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'pysam'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[8], line 15\u001b[0m\n\u001b[0;32m     13\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mre\u001b[39;00m\n\u001b[0;32m     14\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mnumpy\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mnp\u001b[39;00m\n\u001b[1;32m---> 15\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpysam\u001b[39;00m\n\u001b[0;32m     16\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mpysam\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m VariantFile\n\u001b[0;32m     17\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mjson\u001b[39;00m\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'pysam'"
     ]
    }
   ],
   "source": [
    "# packages\n",
    "\n",
    "import re\n",
    "from tqdm import tqdm\n",
    "from ast import literal_eval\n",
    "import os\n",
    "import gzip\n",
    "from tqdm import tqdm\n",
    "from collections import Counter\n",
    "import os\n",
    "import numpy as np\n",
    "import itertools\n",
    "import re\n",
    "import numpy as np\n",
    "import pysam\n",
    "from pysam import VariantFile\n",
    "import json\n",
    "from statistics import mean\n",
    "import statistics\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "import time\n",
    "import tracemalloc\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "25833c57-6d57-438d-a904-29b2f274d05b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'pysam'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[6], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpysam\u001b[39;00m\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'pysam'"
     ]
    }
   ],
   "source": [
    "import pysam"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eba7c00-8cfe-456e-ba9d-2503382f3183",
   "metadata": {},
   "source": [
    "## Step1 - Preparing alignments to realign. \n",
    "1.Trimming low quality ends. -----------  2.Trimming read which exceeds insert size threshold, STR to read ref."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "eb32fb4f-95f2-4f8d-8c4f-402df31c9292",
   "metadata": {},
   "outputs": [],
   "source": [
    "# left alignment. chr1:1762760-1762791\n",
    "\n",
    "# Do you know what is this cell doing ? : 1.\n",
    "\n",
    "class realign:\n",
    "\n",
    "    # default running def and initilization of variables.\n",
    "    def __init__(self, duplicated_read, read_reference_start, read_reference_end, query_seq, query_id, qualities_, bases_, cigar_len, cigar_type, end_pos_, pos_):\n",
    "        \n",
    "        self.duplicated_read       = duplicated_read\n",
    "        self.read_reference_start  = read_reference_start\n",
    "        self.read_reference_end    = read_reference_end\n",
    "        self.query_seq             = query_seq\n",
    "        self.query_id              = query_id\n",
    "        self.qualities_            = qualities_\n",
    "        self.bases_                = bases_\n",
    "        self.cigar_len             = cigar_len\n",
    "        self.cigar_type            = cigar_type\n",
    "        self.end_pos_              = end_pos_\n",
    "        self.pos_                  = pos_\n",
    "        # self.recursion             = recursion\n",
    "        # self.Trimmed_low_quals_list= Trimmed_low_quals_list\n",
    "    \n",
    "    # \n",
    "    def TrimAlignment(self, min_read_start, max_read_stop, min_base_qual, Trim_qual_bases):\n",
    "\n",
    "            \n",
    "        ltrim     = 0\n",
    "        start_pos = self.pos_\n",
    "\n",
    "        # As we use the recursion for both \"Ends trimming\" & \"Far away STRs trimming\".\n",
    "        if Trim_qual_bases == True: # when trimming \"low qual bases\"\n",
    "            i = 1\n",
    "        else: i = 0 # this is when \"Trim bases which are far away\"\n",
    "        \n",
    "        # left to right trimming.        \n",
    "        while((start_pos < min_read_start) and (len(self.cigar_type) > i)): #and len(cigar_ops_) > 0) ---note: cigar len is being checked & i assume all cigar lens will be greater than 0, until it is unmapped.\n",
    "            \n",
    "            #check if we should stop trimming b/c the quality score is above the threshold.\n",
    "            qual_above_thresh = False\n",
    "\n",
    "            # Trim the 40bp either sides of the STR.\n",
    "            if Trim_qual_bases == True:    \n",
    "                if self.cigar_type[0] in ['M', '=', 'X', 'I', 'S']: # do notin if 'D', 'H', cigar_ops[ltrim] was used \n",
    "                         qual_above_thresh = (self.qualities_[ltrim] > min_base_qual)\n",
    "                         \n",
    "            # here we check if qual is above threshold\n",
    "            if qual_above_thresh == True:\n",
    "                break\n",
    "             \n",
    "            # here we increament the start and ltrim positions. # do notin if 'H'\n",
    "            if self.cigar_type[0] in ['M', '=', 'X']:\n",
    "                ltrim += 1\n",
    "                start_pos += 1\n",
    "            elif self.cigar_type[0] in ['D']: start_pos += 1\n",
    "            elif self.cigar_type[0] in ['I', 'S']: ltrim += 1\n",
    "\n",
    "            # removing the bases using CIGAR string.\n",
    "            if self.cigar_len[0] == 1:\n",
    "                self.cigar_len.remove(cigar_len[0])\n",
    "                self.cigar_type.remove(cigar_type[0])\n",
    "            else: self.cigar_len[0] -= 1\n",
    "\n",
    "            # removing the base from the read sequence, ( only I, S, M can be removed, D is not present in the read.)\n",
    "            \n",
    "\n",
    "        # ***************************************************************************************************************************************************\n",
    "        # Right to left trimming.\n",
    "        rtrim = 0\n",
    "        qual_string_len = int(len(self.qualities_)) - 1\n",
    "        rtrim_index     = -rtrim-1\n",
    "        end_pos         = self.end_pos_\n",
    "\n",
    "        while((end_pos > max_read_stop) and (len(self.cigar_type) > i)):\n",
    "\n",
    "\n",
    "            self.qual_above_thresh = False\n",
    "\n",
    "            # Trim the 40bp either sides of the STR.\n",
    "            if Trim_qual_bases == True:    \n",
    "                #\n",
    "                if  self.cigar_type[0] in ['M', '=', 'X', 'I', 'S']:\n",
    "                    self.qual_above_thresh = (self.qualities_[rtrim] > min_base_qual)\n",
    "\n",
    "            \n",
    "            # here we check if qual is above threshold\n",
    "            if self.qual_above_thresh == True: break\n",
    "\n",
    "            # here we increament the start and rtrim positions.\n",
    "            if self.cigar_type[0] in ['M', '=', 'X']:\n",
    "                rtrim -= 1\n",
    "                end_pos -= 1\n",
    "            elif self.cigar_type[0] in ['D']: end_pos -= 1\n",
    "            elif self.cigar_type[0] in ['I', 'S']: rtrim -= 1\n",
    "\n",
    "            # removing the bases using CIGAR string.\n",
    "            if self.cigar_len[-1] == 1:\n",
    "                self.cigar_len.remove(self.cigar_len[-1])\n",
    "            else: self.cigar_len[-1] -= 1\n",
    "                \n",
    "                        \n",
    "\n",
    "        # \n",
    "        assert(ltrim + rtrim <= len(self.bases_)), \"Checking if all bases are trimmed\"\n",
    "        self.bases_     = self.bases_[ltrim : len(self.bases_)+rtrim]\n",
    "        self.qualities_ = self.qualities_\n",
    "        length_         = ltrim - rtrim\n",
    "        self.pos_       = start_pos\n",
    "        self.end_pos_   = end_pos\n",
    "\n",
    "\n",
    "        return self.bases_, start_pos, end_pos, self.cigar_len, self.cigar_type\n",
    "\n",
    "    \n",
    "    # Function pases the min base qual and returns list with deleted low base qualities.\n",
    "    def TrimLowQualityEnds(self, min_base_qual):\n",
    "\n",
    "        return realign.TrimAlignment(self, self.end_pos_+1, self.pos_-1, min_base_qual, True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ca3f2ddd-c351-489e-8044-b87b77fa74c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# realign the reads with left normalization concept. (AlignmentOps)\n",
    "'''\n",
    "Realign read to reference region using left alignment variant.\n",
    "Store the new alignment info using the provided alignment reference.\n",
    "Converts the bases to their upper case variants.\n",
    "'''\n",
    "# fetch part of the sequence \"sequence 1\" from position 0 to 3\n",
    "from pysam import FastaFile\n",
    "\n",
    "# Do you know what is this cell doing ? : 1.\n",
    "\n",
    "def Generate_realign_seq(chrom, ref_start, ref_end):\n",
    "\n",
    "    fasta = \"../../../../malini/repeats/reference/Homo_sapiens_assembly38.fasta\"\n",
    "    sequences_object   = FastaFile(fasta)\n",
    "    # chrom_size         = len(sequences_object.fetch(chrom)\n",
    "    reference_sequence = sequences_object.fetch(chrom, ref_start, ref_end)\n",
    "\n",
    "    # return\n",
    "    return reference_sequence\n",
    "\n",
    "\n",
    "# function which creates the candidate seqs \n",
    "# def Gen_candidate_seqs():\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "4317d732-eedc-4186-a562-219e3be77da7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to generate the Cigar string from given, refn seqn and read seqn. (Global alignment)\n",
    "\n",
    "def Generate_cigar(reference_seq, query_seq):\n",
    "\n",
    "    # Extract CIGAR string from the first alignment\n",
    "    query_lindex  = len(query_seq) - len(query_seq.lstrip('-')) # query indexs helps us to remove \"leading spaces\" & \"trailing spaces\"\n",
    "    query_rindex  = len(query_seq) - len(query_seq.rstrip('-')) \n",
    "    \n",
    "    \n",
    "    # appending the cigar ops in order.\n",
    "    cigar_ops = []\n",
    "    \n",
    "    for pos in range(0+query_lindex, len(query_seq)-query_rindex):\n",
    "    \n",
    "        # refn and query seq each base to check its type.\n",
    "        rbase = reference_seq[pos]\n",
    "        qbase = query_seq[pos]\n",
    "        \n",
    "        # checking \"Macthes\": represents \"M\"\n",
    "        if rbase == qbase :\n",
    "            cigar_ops.append('M')\n",
    "    \n",
    "        # checking \"Insertion\": \"I\"\n",
    "        elif rbase == '-':\n",
    "            cigar_ops.append('I')\n",
    "            \n",
    "        # checking \"Deletion\": \"D\"\n",
    "        elif qbase == '-':\n",
    "            cigar_ops.append('D')\n",
    "        \n",
    "        # checking \"Sub\": \"X\"\n",
    "        elif rbase != qbase:\n",
    "            cigar_ops.append('X')\n",
    "            \n",
    "    \n",
    "    # list cigar ops into joint cigar string.\n",
    "    cigar_ops_values = {'M':0, 'I':0, 'D':0, 'X':0}\n",
    "    cigar_type     = []\n",
    "    cigar_num      = []\n",
    "    \n",
    "    current_ops = cigar_ops[0]\n",
    "    for i,ops in enumerate(cigar_ops):\n",
    "    \n",
    "        # keeps adding the type ops onl if consicutive.\n",
    "        if ops == current_ops:\n",
    "            cigar_ops_values[ops] += 1\n",
    "        \n",
    "        else: # here we realised order is changed with new ops.\n",
    "            cigar_type.append(current_ops) # append the current info\n",
    "            cigar_num.append(cigar_ops_values[current_ops])\n",
    "            cigar_ops_values.update({}.fromkeys(cigar_ops_values,0)) # its necessary to keep the track of cigar ops so we clear dict so we only update order wise.\n",
    "            cigar_ops_values[ops] += 1\n",
    "            current_ops = ops\n",
    "            \n",
    "        if ops == current_ops and len(cigar_ops)-1 == int(i):\n",
    "            cigar_type.append(current_ops)\n",
    "            cigar_num.append(cigar_ops_values[current_ops])\n",
    "        \n",
    "\n",
    "    # return the cigar_type and cigar_num bases.\n",
    "    query_seq1 = query_seq[0+query_lindex : len(query_seq)-query_rindex]\n",
    "    return [cigar_type, cigar_num, query_seq1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "0f059020-03e7-455d-808a-6c4d9b448231",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to extract the seqn from realigned result.\n",
    "\n",
    "def extract_sequence(aln_refn, aln_query, aln_start, aln_end, region_start, region_end, seq):\n",
    "\n",
    "    # we pass the aligned query seqn and reference sequence to function Generate Cigar.\n",
    "    # test print\n",
    "    CIGAR_return = Generate_cigar(aln_refn, aln_query)\n",
    "\n",
    "    cigar_type   = CIGAR_return[0]\n",
    "    cigar_num    = CIGAR_return[1]\n",
    "    query_seq    = CIGAR_return[2]\n",
    "\n",
    "    # check aln start and end enclosing the STR region.\n",
    "    if aln_start >= region_start or aln_end <= region_end: return False\n",
    "\n",
    "    align_index = 0 # index of align type\n",
    "    char_index  = 0 # index of current base in current CIGAR element\n",
    "    pos         = aln_start\n",
    "    \n",
    "    # Extract region sequence if fully spanned by align.\n",
    "    reg_seq = \"\"\n",
    "    cigar_iter = 0\n",
    "\n",
    "    #test print\n",
    "    # print(\"E Seq\", region_start, region_end, aln_start, aln_end, pos, aln_query)\n",
    "    while ( cigar_iter <= len(cigar_num)):\n",
    "        if char_index == cigar_num[cigar_iter]:\n",
    "            # print(\"char_index == cigar_num\", char_index, cigar_num[cigar_iter])\n",
    "            cigar_iter += 1\n",
    "            char_index = 0\n",
    "        \n",
    "        elif pos > region_end:\n",
    "            if reg_seq == \"\":\n",
    "                seq = \"\"\n",
    "            else:\n",
    "                seq = reg_seq\n",
    "            # return\n",
    "            # print(\"pos>region_end\", seq, reg_seq)\n",
    "            return seq  # else it will be an infinite loop\n",
    "        \n",
    "        elif pos == region_end:\n",
    "            # print(\"pos == region_end\", pos, region_end)\n",
    "            if cigar_type[cigar_iter] == 'I':\n",
    "                reg_seq += query_seq[align_index : align_index + int(cigar_num[cigar_iter])]\n",
    "                align_index += int(cigar_num[i])\n",
    "                char_index = 0\n",
    "                cigar_iter += 1\n",
    "            else:\n",
    "                if reg_seq == \"\":\n",
    "                    seq = \"\"\n",
    "                else:\n",
    "                    seq = reg_seq\n",
    "                # return\n",
    "                # print(\"pos==region_end\", seq)\n",
    "                return seq # else it will be infinite loop\n",
    "\n",
    "        elif pos >= region_start:\n",
    "            num_bases = min(region_end-pos, int(cigar_num[cigar_iter])-char_index)\n",
    "            # print(\"pos>=region_start----\", region_start, pos)\n",
    "            if cigar_type[cigar_iter] == 'I':\n",
    "                num_bases = cigar_num[cigar_iter]\n",
    "                reg_seq += query_seq[align_index : align_index + num_bases]\n",
    "                # print(\"pos >= region_start, I\", reg_seq, align_index,\":\",num_bases, \"---------\", pos)\n",
    "            elif cigar_type[cigar_iter] in ['M', 'X']:\n",
    "                reg_seq += query_seq[align_index : align_index + num_bases]\n",
    "                # print(\"pos >= region_start, -------------- M X\", num_bases, align_index,\":\",num_bases, \"---------\", pos, query_seq[align_index : align_index+num_bases])\n",
    "                pos += num_bases\n",
    "            elif cigar_type[cigar_iter] == 'D':\n",
    "                pos += num_bases\n",
    "                # print(\"D---\", pos)\n",
    "            else: print(\"Invalid CIGAR char in extract reg seq\")\n",
    "\n",
    "            align_index += num_bases\n",
    "            char_index  += num_bases\n",
    "\n",
    "        else:\n",
    "            num_bases = 0\n",
    "            if cigar_type[cigar_iter] == 'I':\n",
    "                num_bases = cigar_num[cigar_iter] - char_index\n",
    "            else:\n",
    "                num_bases = min(region_start - pos, cigar_num[cigar_iter] - char_index)\n",
    "                # print(\"else---\", num_bases, region_start, pos, cigar_num[cigar_iter], char_index)\n",
    "                pos += num_bases\n",
    "                # print(\"pos increament in else--\", pos)\n",
    "\n",
    "            \n",
    "            align_index += num_bases\n",
    "            char_index  += num_bases\n",
    "\n",
    "    # return\n",
    "    return False\n",
    "    print(\"Logical error in extract_sequence\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "1f0456f2-731f-41f6-8285-f167bbf8a52c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here we trim identical regions.\n",
    "\n",
    "def trim(ideal_min_length, region_start, region_end, sorted_sequences_list):\n",
    "\n",
    "    min_len = int(len(sorted_sequences_list[1]))\n",
    "    assert(min_len >= ideal_min_length), \"min allele len is <= ideal_min_length\"\n",
    "\n",
    "    # removing identical haplotype sequences\n",
    "    \n",
    "        \n",
    "        \n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "2f4878db-ca1a-414e-97ed-aecaab222adf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function : Main function for this - add_haplotype_block() -link- get_aln_bounds() -link- gen_candidae_seqs().  \n",
    "\n",
    "# Function to extract candidate STR sequences - using some padding to ensure indels near STR ends are included.\n",
    "#                                              1.Ensure that we don't exceed the chr boundaries\n",
    "#                                              2.Extract the alignment boundaries\n",
    "#                                              3.Ensure that we have some spanning alignments to work with\n",
    "#                                              4.Extend each VCF allele by padding size\n",
    "#                                              5.Extract condidate STR seqn (using some padding to ensure indels near STR ends are included)\n",
    "#                                              6.Ensure that the new haplotype block won't overlap with previous blocks\n",
    "#                                              7.Add the new haplotype block\n",
    "\n",
    "#\n",
    "MIN_FRAC_READS           =  0.05\n",
    "MIN_FRAC_SAMPLES         =  0.05\n",
    "MIN_FRAC_STRONG_SAMPLE   =  0.2\n",
    "MIN_READS_STRONG_SAMPLE  =  2\n",
    "MIN_STRONG_SAMPLES       =  1\n",
    "MIN_BLOCK_SPACING        =  10\n",
    "REF_FLANK_LEN            =  35\n",
    "\n",
    "list_start_end_cords1 = []\n",
    "\n",
    "\n",
    "def get_aln_bounds(list_start_end_cords):\n",
    "\n",
    "    max_aln_stop   =  2147483647\n",
    "    min_aln_start  = -2147483647\n",
    "\n",
    "    list_start_end_cords1 = list_start_end_cords\n",
    "    # determining the min and max alignment boundaries.\n",
    "    for i in range(int(len(list_start_end_cords)/4)):\n",
    "        \n",
    "        min_aln_start = min(min_aln_start, list_start_end_cords[i*4])\n",
    "        max_aln_stop  = max(max_aln_stop, list_start_end_cords[i*4+1])\n",
    " \n",
    "    # in pool of reads, we track the min_start and max_end of reads.\n",
    "    return [min_aln_start, max_aln_stop]\n",
    "    \n",
    "\n",
    "\n",
    "# gen_candidate_seq checks for the valid haplotyes by referring the num of reads and samples.\n",
    "def gen_candidate_seq(ref_seq, alignments_start_end, ideal_min_length, list_start_end_cords, region_start, region_end):\n",
    "\n",
    "    sequences = []\n",
    "    assert(len(sequences) == 0), \"sequence list is not empty\"\n",
    "\n",
    "    sample_counts = {} # [seq][count]\n",
    "    read_counts   = [] # [seq][count]\n",
    "    must_inc      = []\n",
    "    total_reads   = 0\n",
    "    total_samples = 0\n",
    "\n",
    "    counts = []\n",
    "    sample_reads = 0\n",
    "    ref_index = -1\n",
    "    # Determines the number of reads and number of samples spporting each allele\n",
    "    for i in range(int(len(alignments_start_end)/4)):\n",
    "        \n",
    "        # print(alignments_start_end[i*4+3], alignments_start_end[i*4+2], len(alignments_start_end))\n",
    "        aln_start = alignments_start_end[i*4]\n",
    "        aln_end   = alignments_start_end[i*4+1]\n",
    "        subseq    = \"\"\n",
    "        # print(\"Es--\", alignments_start_end[i*4+2])\n",
    "        Extracted_sequence = extract_sequence(alignments_start_end[i*4+3], alignments_start_end[i*4+2], aln_start, aln_end, region_start, region_end, subseq)\n",
    "        # print(\"subseq--\", Extracted_sequence)\n",
    "        \n",
    "        if Extracted_sequence != False:\n",
    "\n",
    "            read_counts.append(Extracted_sequence)\n",
    "            counts.append(Extracted_sequence)\n",
    "            total_reads  += 1\n",
    "            sample_reads += 1\n",
    "\n",
    "\n",
    "    # Identify alleles strongly supported by sample\n",
    "    read_counts_dict = Counter(read_counts)    \n",
    "    counts_dict = Counter(counts)\n",
    "\n",
    "    for seqn in counts_dict: # keys as Extraced seqn and its freqn as values.\n",
    "        \n",
    "        # criteria 1\n",
    "        if counts_dict[seqn] >= MIN_READS_STRONG_SAMPLE and counts_dict[seqn] >= MIN_FRAC_STRONG_SAMPLE * sample_reads :\n",
    "            must_inc.append(seqn)\n",
    "        \n",
    "        # criteria 2        \n",
    "        if seqn in sample_counts:\n",
    "            sample_counts[seqn] += counts_dict[seqn] * 1.0 / sample_reads\n",
    "        else: \n",
    "            sample_counts[seqn] = 0\n",
    "            sample_counts[seqn] += counts_dict[seqn] * 1.0 / sample_reads\n",
    "\n",
    "    \n",
    "    #\n",
    "    if sample_reads > 0:\n",
    "        total_samples += 1\n",
    "\n",
    "    # Add alleles with strong support from a subset of samples.\n",
    "    must_inc_dict = Counter(must_inc)\n",
    "    for sqn in must_inc_dict:\n",
    "\n",
    "        if must_inc_dict[sqn] >= MIN_STRONG_SAMPLES:\n",
    "            sequences.append(sqn)\n",
    "            del sample_counts[sqn]\n",
    "            del read_counts_dict[sqn]\n",
    "            if sqn == ref_seq:\n",
    "                ref_index = len(sequences) - 1\n",
    "\n",
    "    \n",
    "    # Identify additional alleles satisfying thresholds\n",
    "    for sqn in sample_counts:\n",
    "\n",
    "        if sample_counts[sqn] > MIN_FRAC_SAMPLES*total_samples or read_counts_dict[sqn] > MIN_FRAC_READS*total_reads:\n",
    "            sequences.append(sqn)\n",
    "            if ref_index == -1 and sqn == ref_seq:\n",
    "                ref_index = len(sequences) - 1\n",
    "\n",
    "    # Arrange reference sequence as first element\n",
    "    if ref_index == -1:\n",
    "        sequences.insert(0, ref_seq)\n",
    "    else:\n",
    "        sequences.insert(ref_index, sequences[0])\n",
    "        sequences.insert(0, ref_seq)\n",
    "        \n",
    "\n",
    "    # Sort regions by length and then by sequence, note: currently only sorting by the len.\n",
    "    sorted_sequences = sorted(sequences[1:], key=lambda x: len(x))\n",
    "    \n",
    "    sorted_sequences_list = sequences[0:1]\n",
    "    sorted_sequences_list.extend(sorted_sequences)\n",
    "\n",
    "    # Clip identical regions\n",
    "    # print(\"sequences--:\", sorted_sequences_list)\n",
    "    trim(ideal_min_length, region_start, region_end, sorted_sequences_list)\n",
    "   \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def add_haplotype_block(add_haplotype_block_values, alignments_start_end, start, end, LEFT_PAD, RIGHT_PAD, REF_FLANK_LEN):\n",
    "\n",
    "    # check 1:\n",
    "    # check if haplotype is passed the criteria, from the function build_haplotype()\n",
    "    # code\n",
    "\n",
    "    # check 2:\n",
    "    # Ensure that we don't exceed the chromosome bounds\n",
    "    chrom_size = 285000000 #temporarily using chr1 value for one region.\n",
    "\n",
    "    if (start < REF_FLANK_LEN + LEFT_PAD   or   end + REF_FLANK_LEN + RIGHT_PAD > chrom_size):\n",
    "        print(\"Haplotype blocks are too near to the chromosome ends\")\n",
    "        return False\n",
    "\n",
    "    min_aln_start = add_haplotype_block_values[0]\n",
    "    max_aln_stop  = add_haplotype_block_values[1]\n",
    "    # Ensuring that we have some spanning alignmnets to work with\n",
    "    region_start  = start - LEFT_PAD\n",
    "    region_end    = end   + RIGHT_PAD\n",
    "    refn_seq      = Generate_realign_seq(\"chr1\", region_start, region_end) # check and pull the req seq from already pulled seq, up and down 74bp\n",
    "    if (min_aln_start + 5 >= region_start or max_aln_stop - 5 <= region_end):\n",
    "        print(\"No spanning alignment\")\n",
    "        return False\n",
    "\n",
    "    # Extend each vcf allele by padding size, no idea why is this snipet being used ******* why\n",
    "    # padded_vcf_alleles = []\n",
    "\n",
    "\n",
    "    # Extract candidate STR seqn ( using padding to ensure indels near STR ends are included )\n",
    "    ideal_min_length = 3*5 # test case\n",
    "    gen_candidate_seq(refn_seq, alignments_start_end, ideal_min_length, add_haplotype_block_values, region_start, region_end)\n",
    "    \n",
    "     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "3f8d8e1c-dc0a-4515-b266-ca11e6d7be41",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'CTCAAAATGACTGAATGAATGAATGAATGACTGAATAA', 'CTCAAAATGAATGAATGAATGAATGAATGAATGAATAA', 'CTCAAAATGAATGAATGAATGAATGAATGAATGAATGAATAA', 'CTCAAAATGAATGACTGAATGAATGAATGAATGAATAA'} trim\n"
     ]
    }
   ],
   "source": [
    "# parsing the .bam file and fetching the reads corresponding to the repeat regions and getting their informative reads.\n",
    "\n",
    "# Do you know what is this cell doing ? : 1.\n",
    "\n",
    "Pysam_Read_directory = \"../../../../malini/repeats/Short_Read_Genotyping/\" \n",
    "single_Bamfile_id       = \"LK66312722JN\"\n",
    "\n",
    "samfile = pysam.AlignmentFile(f'{Pysam_Read_directory}{single_Bamfile_id}.bam', \"rc\")\n",
    "\n",
    "#\n",
    "chrom      = \"chr1\" \n",
    "start      = 1762760 - 1 \n",
    "end        = 1762791\n",
    "repeat_len = int(end)-int(start)\n",
    "LEFT_PAD       = 5\n",
    "RIGHT_PAD      = 5\n",
    "REF_FLANK_LEN  = 35\n",
    "\n",
    "# Print the alignments and their scores.\n",
    "alignments_start_end = []\n",
    "\n",
    "\n",
    "for read in samfile.fetch(chrom,start, end):\n",
    "\n",
    "    # \n",
    "    if read.is_supplementary or read.is_duplicate : continue\n",
    "    if read.reference_start > start or read.reference_end < end: continue # @@@@@@@@ check if this line is required\n",
    "\n",
    "    # query : start, end, min_base_quality. \n",
    "    query_start      = read.reference_start\n",
    "    query_end        = read.reference_end\n",
    "    min_base_qual    = 12 # ~ = 12\n",
    "    \n",
    "    # cigar string, (type & len)\n",
    "    cigar_string = read.cigarstring\n",
    "    cigar_len_t  = re.findall(r'[0-9]+', cigar_string)\n",
    "    cigar_len    = []\n",
    "    temp         = [cigar_len.append(int(x)) for x in cigar_len_t] \n",
    "    cigar_type   = re.findall(r'[a-zA-Z]+', cigar_string)\n",
    "\n",
    "    # # Ignoring whos cigars only have, 'H' or 'S' with 'M'.\n",
    "    # if cigar_type.count('H') != 0 or cigar_type.count('S') != 0:\n",
    "    #     if cigar_type.count('I') == 0 or cigar_type.count('D') == 0 or cigar_type.count('')\n",
    "    \n",
    "    # qualities_, bases_ and list of cigar operations\n",
    "    qualities_ = list(read.query_qualities)\n",
    "    bases_     = list(read.query_sequence)\n",
    "\n",
    "    \n",
    "    # Class object, obj\n",
    "    # tp\n",
    "    obj = realign(read.is_duplicate, read.reference_start, read.reference_end, read.seq, read.query_name, qualities_, bases_, cigar_len, cigar_type, query_end, query_start)\n",
    "    Trimmed_ends = obj.TrimLowQualityEnds(min_base_qual)\n",
    "\n",
    "        \n",
    "    # Trimming the both repeat ends which are far away, 40bp upstream and down stream\n",
    "    max_read_stop = end +  40\n",
    "    if start > 40: min_read_start = start - 40\n",
    "    else: min_read_start = 1\n",
    "\n",
    "\n",
    "    # Trimmed too far away sequence from the STR, but threshold is arbitary.\n",
    "    Trimmed_bases   = obj.TrimAlignment(min_read_start, max_read_stop, 12, False)\n",
    "    Seq_realign     = \"\".join(Trimmed_bases[0])\n",
    "    realign_start   = Trimmed_bases[1]\n",
    "    realign_end     = Trimmed_bases[2]\n",
    "    realign_cigar_l = Trimmed_bases[3]\n",
    "    realign_cigar_t = Trimmed_bases[4] \n",
    " \n",
    "    \n",
    "    # Generating ref seq. *** check for if refern position is 0 and size of the chr for ref_stop.    \n",
    "    Align_window_width = 75\n",
    "    ref_start = realign_start - Align_window_width - 1\n",
    "    ref_stop  = realign_end   + Align_window_width\n",
    "\n",
    "    # Select only those alignments marked as good for haplotype generation.\n",
    "    success = True\n",
    "    \n",
    "    \n",
    "    # Function to realign the seqn using left align technique.\n",
    "    # 1. Global alignment from Trimmed read and Expanded STR region.\n",
    "    ref_seq  = Generate_realign_seq(chrom, ref_start, ref_stop)\n",
    "    if len(cigar_type) > 1:\n",
    "        # Global alignment\n",
    "        read_seq = Seq_realign\n",
    "    \n",
    "    \n",
    "        # Perform global alignment using Needleman-Wunsch algorithm\n",
    "        alignments = pairwise2.align.globalms(ref_seq, read_seq, 2, -2, -5, -0.125, one_alignment_only=True) #(ref_eq, read_seq, match_score, mismatch_penalty, open_gap_penalty, extend_gap_penalty)\n",
    "    \n",
    "        new_aligned_seq = alignments[0][1]\n",
    "        alignments_start_end.extend([realign_start, realign_end, new_aligned_seq, alignments[0][0]])\n",
    "\n",
    "    elif len(cigar_type) == 1: # cross check ***********\n",
    "\n",
    "        # here we check the flanks of the STR region to ensure indels or sub.\n",
    "        check_str_region = ref_seq[realign_start - ref_start : realign_end - ref_stop]\n",
    "\n",
    "        # if we notice var in flanks of read, then ignore it.\n",
    "        if check_str_region == Seq_realign:\n",
    "            alignments_start_end.extend([realign_start, realign_end, check_str_region, ref_seq])        \n",
    "\n",
    "        # INCLUDE CODE:**********\n",
    "        # here we check whether to add \"sameAsref\" with flank indel/sub  or  \"sameAsref\" without flank indel/sub.\n",
    "        # code:\n",
    "        # flank-var should be more than 20% of reads.\n",
    "        # if flank-var >= 20% : print(\"likely ambigous region\")\n",
    "\n",
    "\n",
    "# print(\"len of align-:\", len(alignments_start_end)/4)\n",
    "\n",
    "# passing the arguments as list of alignment start end.\n",
    "add_haplotype_block_values = get_aln_bounds(alignments_start_end)\n",
    "add_haplotype_block(add_haplotype_block_values, alignments_start_end, start, end, LEFT_PAD, RIGHT_PAD, REF_FLANK_LEN)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b32a975e-4667-4894-a07c-71b570715371",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ad93390-e894-4a95-a63c-5f17c7d2c39b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e676011-e66d-43c0-981b-cb5bbb6816ed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "882b84ab-8f45-4cd5-8a9d-fc344f514e6b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4af9495-b548-4f49-b9e4-6724715b5022",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8be84a2-ddb7-4e12-83e1-4114e507d015",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e23eaffd-d37c-4060-97fe-eef8771b2e0c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bdaddad-e7c9-4eaa-b607-69b7369b2a67",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93799835-8fd8-497b-a327-7900aed8025a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6725c88c-8b66-4079-bf0c-03db26863f8a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3aa4e971-fca2-4392-94dc-3b2a2e75c85d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
